### PR TITLE
Provide a way to set a gRPC's ClientInterceptor using GrpcClientOptions

### LINKED ISSUE #3964: Provide a way to set a gRPC's `ClientInterceptor` using `GrpcClientOptions`
The official client factory (`Clients`) can not take `ClientInterceptor`s from its builder method.
Users are able to add `ClientInterceptor`s using upstream gRPC's API after a client stub is created. 
```java
HelloServiceStub helloService = Clients.newClient(grpcUri, HelloServiceStub.class);
helloService.withInterceptors(myInterceptor1, myInterceptor2);
```
I think it would be convenient to fluently add `ClientInterceptor` using `GrpcClientOptions`.
For example:
```java
// Don't need to mutate the created client stub anymore
final HelloServiceStub helloService = 
    Clients.builder(grpcUri)
           .option(GrpcClientOptions.INTERCEPTORS.newValue(List.of(myInterceptor1, myInterceptor2))
           .build(HelloServiceStub.class);
```

### PR DESCRIPTION
**Pull request by ta7uw**
Motivation:

Users are able to add `ClientInterceptors` using upstream gRPC's API after a client stub is created.
it would be convenient to fluently add `ClientInterceptor` using `GrpcClientOptions`.

Modifications:

- Add `INTERCEPTORS` to `GrpcClientOptions` as a static field.
- Create a new Channel that will call interceptors before starting a call on the given `ArmeriaChannel` in `GrpcClientFactory#newClient`.

Result:

- Closes #3964
- Set gRPC's `ClientInterceptor` using `GrpcClientOptions`

```java
// Don't need to mutate the created client stub anymore
final HelloServiceStub helloService = 
    Clients.builder(grpcUri)
           .option(GrpcClientOptions.INTERCEPTORS.newValue(List.of(myInterceptor1, myInterceptor2))
           .build(HelloServiceStub.class);
```

### CODE REVIEW

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 1]
```diff
@@ -129,5 +132,13 @@
             GRPC_CLIENT_STUB_FACTORY = ClientOption.define("GRPC_CLIENT_STUB_FACTORY",
                                                            NullGrpcClientStubFactory.INSTANCE);
 
+    /**
+     * Sets the {@link ClientInterceptor}s to the gRPC client stub.
+     * The last interceptor will have its ClientInterceptor.interceptCall called first.
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:137]:
```suggestion
     * The last interceptor will have its {@code ClientInterceptor.interceptCall()} called first.
```

File: grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcClientFactory.java [hunk 2]
```diff
@@ -133,15 +135,15 @@ public Object newClient(ClientBuilderParams params) {
...
+        final Channel channel = ClientInterceptors.intercept(new ArmeriaChannel(
                 newParams,
                 httpClient,
                 meterRegistry(),
                 scheme.sessionProtocol(),
                 serializationFormat,
                 jsonMarshaller,
-                simpleMethodNames);
-
+                simpleMethodNames), interceptors);
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcClientFactory.java:146]: Could we apply `ClientInterceptors.intercept` only when `interceptors` is non empty?
```java
if (!interceptors.isEmpty()) {
    ...
}
```

**Review by ikhoon — APPROVED** [review]
This is so quick! Thanks, @ta7uw!

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 3]
```diff
@@ -129,5 +132,13 @@
             GRPC_CLIENT_STUB_FACTORY = ClientOption.define("GRPC_CLIENT_STUB_FACTORY",
                                                            NullGrpcClientStubFactory.INSTANCE);
 
+    /**
+     * Sets the {@link ClientInterceptor}s to the gRPC client stub.
+     * The last interceptor will have its {@code ClientInterceptor.interceptCall()} called first.
+     */
+    public static final ClientOption<List<ClientInterceptor>>
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:139]: nit:
```suggestion
    public static final ClientOption<List<? extends ClientInterceptor>>
```

**Review by minwoox — APPROVED** [review]
Just one nit. Thanks a lot. 😄

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 4]
```diff
@@ -129,5 +132,13 @@
             GRPC_CLIENT_STUB_FACTORY = ClientOption.define("GRPC_CLIENT_STUB_FACTORY",
                                                            NullGrpcClientStubFactory.INSTANCE);
 
+    /**
+     * Sets the {@link ClientInterceptor}s to the gRPC client stub.
+     * The last interceptor will have its {@code ClientInterceptor.interceptCall()} called first.
```
Comment (jrhee17) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:137]: nit; following our javadocs for decorators (I think the suggested javadoc makes the order more clear)

https://github.com/line/armeria/blob/baaf023af655c07ab177a2add5e24252e267673c/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java#L1056

```suggestion
     * The specified interceptor(s) is/are executed in reverse order.
```

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 5]
```diff
@@ -129,5 +132,13 @@
...
+     */
+    public static final ClientOption<List<ClientInterceptor>>
+            INTERCEPTORS = ClientOption.define("GRPC_CLIENT_INTERCEPTORS",
```
Comment (jrhee17) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:140]: nit; no need to line wrap
```suggestion
            INTERCEPTORS = ClientOption.define("GRPC_CLIENT_INTERCEPTORS", ImmutableList.of());
```

File: grpc/src/test/java/com/linecorp/armeria/internal/client/grpc/CustomGrpcClientFactoryTest.java [hunk 6]
```diff
@@ -78,4 +115,72 @@ public Object newClientStub(Class<?> clientType, Channel channel) {
...
+        final ClientInterceptor firstInterceptor = new ClientInterceptor() {
+            @Override
+            public <I, O> ClientCall<I, O> interceptCall(
+                    MethodDescriptor<I, O> method, CallOptions callOptions, Channel next) {
+                return new SimpleForwardingClientCall<I, O>(next.newCall(method, callOptions)) {
+                    @Override
+                    public void start(Listener<O> responseListener, Metadata headers) {
+                        logger.info("first");
+                        invoked1.getAndIncrement();
+                        super.start(responseListener, headers);
```
Comment (jrhee17) [inline grpc/src/test/java/com/linecorp/armeria/internal/client/grpc/CustomGrpcClientFactoryTest.java:132]: nit; what do you think of removing the logging, and asserting from interceptors?
This way, we can be sure the interceptors are run in the correct order.

ditto for the other interceptors

```suggestion
                        invoked1.getAndIncrement();
                        assertThat(invoked1).hasValue(1);
                        assertThat(invoked2).hasValue(0);
                        assertThat(invoked3).hasValue(0);
                        super.start(responseListener, headers);
```

**Review by jrhee17 — COMMENTED** [review]
Looks really nice! Left some trivial comments 🙇 👍 🙇

File: grpc/src/test/java/com/linecorp/armeria/internal/client/grpc/CustomGrpcClientFactoryTest.java (hunk 6 above)
Comment (ta7uw) [reply grpc/src/test/java/com/linecorp/armeria/internal/client/grpc/CustomGrpcClientFactoryTest.java:132]: It seems good

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 7]
```diff
@@ -129,5 +132,12 @@
             GRPC_CLIENT_STUB_FACTORY = ClientOption.define("GRPC_CLIENT_STUB_FACTORY",
                                                            NullGrpcClientStubFactory.INSTANCE);
 
+    /**
+     * Sets the {@link ClientInterceptor}s to the gRPC client stub.
+     * The specified interceptor(s) is/are executed in reverse order
+     */
+    public static final ClientOption<List<? extends ClientInterceptor>>
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:139]: By the way, `Iterable<? extends ClientInterceptor>` looks better.

File: grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java [hunk 8]
```diff
@@ -129,5 +132,12 @@
             GRPC_CLIENT_STUB_FACTORY = ClientOption.define("GRPC_CLIENT_STUB_FACTORY",
                                                            NullGrpcClientStubFactory.INSTANCE);
 
+    /**
+     * Sets the {@link ClientInterceptor}s to the gRPC client stub.
+     * The specified interceptor(s) is/are executed in reverse order
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/client/grpc/GrpcClientOptions.java:137]: Missing a trailing dot.
```suggestion
     * The specified interceptor(s) is/are executed in reverse order.
```

File: grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcClientFactory.java [hunk 9]
```diff
@@ -133,15 +135,18 @@ public Object newClient(ClientBuilderParams params) {
...
                 newParams,
                 httpClient,
                 meterRegistry(),
                 scheme.sessionProtocol(),
                 serializationFormat,
                 jsonMarshaller,
                 simpleMethodNames);
-
+        final List<? extends ClientInterceptor> interceptors = options.get(GrpcClientOptions.INTERCEPTORS);
+        if (!interceptors.isEmpty()) {
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcClientFactory.java:147]: We can use `Iterables.isEmpty()` if we change `List` into `Iterable`.

**Review by ikhoon — COMMENTED** [review]
Left minor nits.

**Review by jrhee17 — APPROVED** [review]
💯  Looks really nice! Thank you @ta7uw ! 🙇 🥳 🎉

**Comment by minwoox** [comment]
Congratulations, @ta7uw! 🎉 🎉 🎉

**Pull request merged**